src/bindings/fluid_ladspa.c

```diff
@@ -22,1107 +22,1288 @@
+static void activate_plugin(fluid_ladspa_plugin_t *plugin);
+static void deactivate_plugin(fluid_ladspa_plugin_t *plugin);
+static fluid_ladspa_plugin_t *get_plugin_by_id(fluid_ladspa_fx_t *fx, int id);
 
-/* Purpose:
- * Returns the pointer to the shared library loaded using 'LibraryFilename' (if it has been loaded).
- * Return NULL otherwise.
- * If not: Clear Fx and abort.
- */
-void * fluid_LADSPA_RetrieveSharedLibrary(fluid_LADSPA_FxUnit_t* FxUnit, char * LibraryFilename){
-  void * CurrentLib=NULL;
-  int LibCount;
-  for (LibCount=0; LibCount<FxUnit->NumberLibs; LibCount++){
-    assert(FxUnit->ppvPluginLibNames[LibCount]);
-    if (FLUID_STRCMP(FxUnit->ppvPluginLibNames[LibCount],LibraryFilename)==0){
-      CurrentLib=FxUnit->ppvPluginLibs[LibCount];
-    };
-  };
-  return CurrentLib;
-};
+static fluid_ladspa_lib_t *new_fluid_ladspa_lib(fluid_ladspa_fx_t *fx, const char *filename);
+static void delete_fluid_ladspa_lib(fluid_ladspa_lib_t *lib);
+static fluid_ladspa_lib_t *get_ladspa_library(fluid_ladspa_fx_t *fx, const char *filename);
+static int load_plugin_library(fluid_ladspa_lib_t *lib);
+static void unload_plugin_library(fluid_ladspa_lib_t *lib);
 
-/* Purpose:
- * Loads a shared LADSPA library.
- * Return NULL, if failed.
- * TODO: use LADSPA_PATH
- */
-void * fluid_LADSPA_LoadSharedLibrary(fluid_LADSPA_FxUnit_t* FxUnit, char * LibraryFilename){
-  void * LoadedLib;
-  assert(LibraryFilename);
-  LoadedLib=dlopen(LibraryFilename,RTLD_NOW);
-  if (!LoadedLib){
-    return NULL;
-  };
-  FxUnit->ppvPluginLibs[FxUnit->NumberLibs]=LoadedLib;
-  FxUnit->ppvPluginLibNames[FxUnit->NumberLibs]=FLUID_STRDUP(LibraryFilename);
-  FxUnit->NumberLibs++;
-  return LoadedLib;
-};
+static FLUID_INLINE void node_to_buffer(fluid_ladspa_node_t *node, fluid_real_t *buffer);
+static FLUID_INLINE void buffer_to_node(fluid_real_t *buffer, fluid_ladspa_node_t *node);
 
-/* Purpose:
- * Retrieves a descriptor to the plugin labeled 'PluginLabel' from the shared library.
+/**
+ * Creates a new LADSPA effects unit.
+ *
+ * @param synth FluidSynth instance
+ * @return pointer to the new LADSPA effects unit
  */
-const LADSPA_Descriptor * fluid_LADSPA_Retrieve_Plugin_Descriptor(void * CurrentLib, char * PluginLabel){
-  LADSPA_Descriptor_Function pfDescriptorFunction;
-  unsigned long lPluginIndex=0;
-  const LADSPA_Descriptor * psDescriptor;
-  pfDescriptorFunction = (LADSPA_Descriptor_Function)dlsym(CurrentLib,"ladspa_descriptor");
-  while (pfDescriptorFunction(lPluginIndex)){
-    psDescriptor = pfDescriptorFunction(lPluginIndex);
-    if (FLUID_STRCMP(psDescriptor->Label, PluginLabel) == 0){
-      return psDescriptor;
-    };
-    lPluginIndex++;
-  };
-  return NULL;
+fluid_ladspa_fx_t *new_fluid_ladspa_fx(fluid_synth_t *synth)
+{
+    fluid_ladspa_fx_t *fx;
+
+    fx = FLUID_NEW(fluid_ladspa_fx_t);
+    if (fx == NULL)
+    {
+        return NULL;
+    }
+
+    fx->state = FLUID_LADSPA_INACTIVE;
+
+    fx->sample_rate = synth->sample_rate;
```

Comment:
I'm not a fan of shadowing variables. And the synth may actually change the samplerate. I guess this would cause trouble. Couldnt you store the synth instance and use the synth API getters instead?

Reply:
The LADSPA engine can't handle sample rate changes yet, so I thought it would be best to make this more or less obvious by copying that setting on instantiation of the LADSPA fx. But this is only temporary, I will add another commit that introduces proper handling of sample rate changes and will address this comment then. Would that be ok for you?

Reply:
Just to be clear... the old LADSPA engine couldn't handle sample rate changes either. Even worse... the sample rate was hardcoded to 44100.

Reply:
I'm a bit two minded about this. On the one hand I would accept it, on the other "Nothing lasts longer than a temporary..."

Reply:
You are so right :-) Let me rephrase: I will address this before signing off on this pull request.

Reply:
ce79b5f adds sample rate change handling. I still shadow the sample rate in the ladspa structure, because the plugins want it in unsigned long and not float. And also because I think it's better to work with a stable value for the sample rate and only change it in fluid_ladspa_set_sample_rate. What do you think?